vendor/@material-ui/types/index.d.ts

```diff
@@ -71,19 +71,3 @@ export type And<A, B, C = true> = A extends true
       : false
     : false
   : false;
-
-/**
- * @internal
- *
- * check if a type is `{}`
- *
- * 1. false if the given type has any members
- * 2. false if the type is `object` which is the only other type with no members
- *  {} is a top type so e.g. `string extends {}` but not `string extends object`
- * 3. false if the given type is `unknown`
- */
-export type IsEmptyInterface<T> = And<
```

Comment:
I was expecting `CoerceEmptyInterface` and maybe `IsAny`, `And`, and `Or` to go as well.  Are they used for other things?

Reply:
> What scenario are you measuring?

```ts

import { makeStyles } from "@material-ui/styles";

// https://github.com/mui-org/material-ui/issues/19113#issuecomment-573614279

const useStaticStyles = makeStyles({ root: { color: "blue" } });
const useDynamicStyles = makeStyles({
  root: { color: (props: { color: string }) => props.color }
});
function Component() {
  const staticClasses = useStaticStyles(); // No error
  const throwingClasses = useDynamicStyles(); // $ExpectError
  const dynamicClasses = useDynamicStyles({ color: "blue" });
}
```

-- https://github.com/eps1lon/mui-types-perf/blob/master/sources/IsEmptyInterface.test.ts

Reply:
[We can remove `CoerceEmptyInterface`](https://github.com/mui-org/material-ui/pull/19259). The others are used to check if we can call `useStyles` with one or zero arguments. 

I'll cherry-pick https://github.com/mui-org/material-ui/pull/19243 and https://github.com/mui-org/material-ui/pull/19259 first and see if we get a more significant improvement.

Reply:
Did the other commits help?

Reply:
Hard to tell since these were only affecting `withStyles` and `styled`. We did get some slight runtime improvements in our `test_types` job but these are only single runs. Might not have been significant.

Having a large codebase where I can just run `tsc` on would help. The original repo in https://github.com/microsoft/TypeScript/issues/34801 didn't help much because it includes a full webpack build (and therefore noise). 

My guess is that types related to CSSProperties are the issue. These are quite complicated so I didn't look into them much. Will be todays and probably tomorrows agenda since today is a bit busier than usual.

Reply:
There were some repro notes in https://github.com/microsoft/TypeScript/issues/34801.  Would they get you more useful numbers?

Reply:
All I see is "takes forever to report errors in vscode". Without know their vscode environment or how long "forever" is I can't do much. Sorry but I can't find detailed instructions.

Reply:
I'm sure I could reconstruct a repro, but I think this is a bit of a tangent anyway - my local experiments suggest that the CSS properties types are more impactful.